### PR TITLE
Fix #2482 Remove usage of propTypes and replace with FlowTypes.

### LINKED ISSUE #2482: Use FlowTypes for frontend components
I was looking at the `eslint-plugin-react/recommended` rules and noticed in more recent versions of React (15.5.0+) and eslint-plugin-react that they removed `React.PropTypes` with the [`prop-types`](http://npm.im/prop-types) module:

```sh
$ npm run lint

> testpilot@1.0.0 lint /Users/pdehaan/dev/github/mozilla/testpilot
> gulp scripts-lint styles-lint

[15:08:14] Requiring external module babel-register
[15:08:17] Using gulpfile ~/dev/github/mozilla/testpilot/gulpfile.babel.js
[15:08:17] Starting 'scripts-lint'...
[15:08:17] Starting 'styles-lint'...
[15:08:21] 'scripts-lint' errored after 3.4 s
[15:08:21] ESLintError in plugin 'gulp-eslint'
Message:
    React.PropTypes is deprecated since React 15.5.0, use the npm module prop-types instead
Details:
    fileName: /Users/pdehaan/dev/github/mozilla/testpilot/frontend/src/app/components/ExperimentTourDialog.js
    lineNumber: 138
[15:08:22] Finished 'styles-lint' after 4.83 s
```

### PR DESCRIPTION
**Pull request by fzzzy**

### CODE REVIEW

File: frontend/src/app/components/LayoutWrapper.js [hunk 1]
```diff
@@ -13,7 +13,9 @@ const LayoutWrapper = ({ children, flexModifier, helperClass = null }) => {
 
 export default LayoutWrapper;
 
-LayoutWrapper.PropTypes = {
-  flexModifier: PropTypes.string,
-  helperClass: PropTypes.string
-};
+// TODO Port propTypes to FlowTypes for LayoutWrapper
```
Comment (meandavejustice) [inline frontend/src/app/components/LayoutWrapper.js]: noting todo here

Comment (meandavejustice) [inline frontend/src/app/components/LayoutWrapper.js:16]: noting todo here

File: frontend/src/app/components/View.js [hunk 2]
```diff
@@ -105,41 +105,42 @@ export default class View extends React.Component {
...
-
-
 View.defaultProps = {
   centered: false,
   showNewsletterFooter: true,
   showFooter: true,
   showHeader: true,
   spaceBetween: false
 };
+
```
Comment (meandavejustice) [inline frontend/src/app/components/View.js:115]: noting todo

File: frontend/src/app/containers/ExperimentPage.js [hunk 3]
```diff
@@ -31,13 +31,14 @@ export default class ExperimentPage extends React.Component {
...
 
+// TODO Implement FlowTypes for ExperimentPage
 
-ExperimentPage.propTypes = {
-  getExperimentBySlug: React.PropTypes.func,
-  params: React.PropTypes.shape({
-    slug: React.PropTypes.string
-  })
-};
+// ExperimentPage.propTypes = {
```
Comment (meandavejustice) [inline frontend/src/app/containers/ExperimentPage.js:36]: todo

**Review by meandavejustice — CHANGES_REQUESTED** [review]

**Comment by lmorchard** [comment]
Are flow type checks run in CircleCI? I tried this and got an error:
```bash
Les Orchard@Preciso MINGW32 ~/devel/testpilot
$ ./node_modules/.bin/flow frontend
Launching Flow server for C:\Users\Les Orchard\devel\testpilot\frontend
Spawned flow server (pid=27580)
Logs will go to C:\Users\LESORC~1\AppData\Local\Temp\flow\CzCzBUserszBLes OrchardzBdevelzBtestpilotzBfrontend.log
Launching Flow server for C:\Users\Les Orchard\devel\testpilot\frontend
Spawned flow server (pid=3656)
Logs will go to C:\Users\LESORC~1\AppData\Local\Temp\flow\CzCzBUserszBLes OrchardzBdevelzBtestpilotzBfrontend.log
frontend/src/app/components/ExperimentTourDialog.js:33
 33:     const { experiment, isExperimentEnabled } = this.props;
                             ^^^^^^^^^^^^^^^^^^^ property `isExperimentEnabled`. Property not found in
 33:     const { experiment, isExperimentEnabled } = this.props;
                                                     ^^^^^^^^^^ object type


Found 1 error
```

**Comment by lmorchard** [comment]
Looked simple enough to fix, though the error should probably be cause to fail the CircleCI run.
```patch
diff --git a/frontend/src/app/components/ExperimentTourDialog.js b/frontend/src/app/components/ExperimentTourDialog.js
index a5cc5f8c..2d385f72 100644
--- a/frontend/src/app/components/ExperimentTourDialog.js
+++ b/frontend/src/app/components/ExperimentTourDialog.js
@@ -7,6 +7,7 @@ import { experimentL10nId } from '../lib/utils';

 type ExperimentTourDialogProps = {
   experiment: Object,
+  isExperimentEnabled: Function,
   onCancel: Function,
   onComplete: Function,
   sendToGA: Function
```

**Comment by lmorchard** [comment]
Maybe this tweak is enough to make this happen on CircleCI?

```patch
diff --git a/package.json b/package.json
index b9e8801d..28a8d1f3 100644
--- a/package.json
+++ b/package.json
@@ -114,7 +114,7 @@
     "static": "gulp dist-build",
     "docs": "doctoc README.md && doctoc addon/README.md",
     "lint": "gulp scripts-lint styles-lint",
-    "test": "cd frontend && mocha --require babel-register --require test-setup.js --recursive",
+    "test": "cd frontend && flow && mocha --require babel-register --require test-setup.js --recursive",
     "test:ci": "cross-env NODE_ENV=test nyc mocha --recursive --reporter mocha-junit-reporter --require frontend/test-setup.js frontend/test",
     "test:watch": "npm test -- --watch",
     "test:all": "npm run lint && npm run test && npm run flow-coverage && cd addon && npm run lint && npm run coverage && npm run test",
```

**Comment by lmorchard** [comment]
Still soaking my head in Flow, so this is kind of a conceptual question: 

We do a lot of things in the pattern of `<SubComponent {...this.props}>` to pass along psuedo-global context from the `<App>` component. We'd get a warning from dev-mode React at run time if those props were missing something. 

Does Flow catch this for us at validation time? I guess I could try breaking something locally and see what happens.

**Comment by fzzzy** [comment]
@lmorchard I'm surprised that circle wasn't running the flow tests, but it seems to have been missing. Good catch, thanks.

About your question re: `...props`, I do think flow tracks that properly. However, it seems to not understand what is happening with react-redux and the components that get their props directly from the App. That's what I put the TODOs there for. I need to learn how to handle that case.

@meandavejustice The TODOs are there because I don't yet know how to handle those cases. Currently HEAD is broken if you do a fresh checkout because the propTypes warnings turned into errors. I'm not able to update our dependencies until we either remove propTypes or switch to the standalone propTypes module. So, I'd prefer if we get HEAD sorted out so it is not busted, and fix the TODOs later.

**Comment by fzzzy** [comment]
Good, I was able to get flow to fail on circleci.

**Comment by fzzzy** [comment]
I filed https://github.com/mozilla/testpilot/issues/2504 to take care of the TODOs.

**Review by meandavejustice — APPROVED** [review]

**Comment by lmorchard** [comment]
Can we squash stuff like this before merging? (eg. the "letr's see" commits?)

**Comment by fzzzy** [comment]
@lmorchard Yep, sorry about that.

**Comment by fzzzy** [comment]
The one thing I wish is that squashing wouldn't ruin the interleaving of commits and comments. Maybe github will figure out how to fix that some day, though.

**Pull request merged**